### PR TITLE
Remove `getaddrinfo` in `SendDummyPacket()` to address malloc deadlock

### PR DESCRIPTION
**Pull request by Chibin**
`SendDummyPacket` eventually calls `getaddrinfo` (which is a reentrant), however, `getaddrinfo` is not an async-signal-safe function. `getaddrinfo` internally calls `malloc`, which is strongly advised to not do within a signal handler as it may cause deadlocks. Cache the accepted socket information for the listener, so that it can be reused in `SendDummyPacket()`.

The purpose of `SendDummyPacket` seems to be to exit more quickly; it circumvents the polling that happens, which eventually times out after 250ms.

Without `SendDummyPacket()`, there will be multiple test failures since some tests expects the backend connection to terminate almost immediately.

To view all the async-signal-safe functions, please view the signal-safety(7) — Linux manual page.

This PR addresses https://github.com/greenplum-db/gpdb/issues/16081

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/cdb/motion/ic_udpifc.c [hunk 1]
```diff
@@ -1278,6 +1280,9 @@ setupUDPListeningSocket(int *listenerSocketFd, uint16 *listenerPort, int *txFami
 		break;
 	}
 
+	memcpy(&g_listener_addrinfo, addr, sizeof(g_listener_addrinfo));
+	memcpy(&g_listener_sockaddr, addr->ai_addr, sizeof(g_listener_sockaddr));
```
Comment (soumyadeep2007) [inline src/backend/cdb/motion/ic_udpifc.c:1284]: I think we should also perform a memset(0) for these two globals in `TeardownUDPIFCInterconnect_Internal`, as a form of cleanup, just for hygiene.

File: src/backend/cdb/motion/ic_udpifc.c [hunk 2]
```diff
@@ -641,6 +641,9 @@ typedef struct ICStatistics
 /* Statistics for UDP interconnect. */
 static ICStatistics ic_statistics;
 
+static struct addrinfo g_listener_addrinfo;
+static struct sockaddr g_listener_sockaddr;
```
Comment (soumyadeep2007) [inline src/backend/cdb/motion/ic_udpifc.c:645]: s/g_listener/udp_dummy_packet

**Review by soumyadeep2007 — COMMENTED** [review]
I reviewed the PR keeping in mind the current approach. However, there are a few things to consider if we can do differently:

1. Bigger change: Why not use the `UDP_listenerFd` socket to send the packet? It is used for IC communication in general. I'm not sure why we ever needed a dedicated socket to send this message. Unfortunately, the historical commit does not really shed any light into why.

2. Smaller change: Create a dedicated socket during setup like `UDP_listenerFd` (re-using `setupUDPListeningSocket()`) and call it `UDP_dummyFd` or something, use that and close it during teardown? Would eliminate a lot of code that way, and keep 1 global and it is also closer in implementation to the other socket stuff.

Both are worth exploring IMO. Getting to option 1 would be awesome. But option 2 seems safer and since this is going into a stable release and will have to be backported to 6X, I would personally lean towards option 2 (with a FIXME comment: "check if we can use UDP_listenerFd to send the dummy message").

Thoughts?

**Comment by Chibin** [comment]
> I reviewed the PR keeping in mind the current approach. However, there are a few things to consider if we can do differently:
> 
>     1. Bigger change: Why not use the `UDP_listenerFd` socket to send the packet? It is used for IC communication in general. I'm not sure why we ever needed a dedicated socket to send this message. Unfortunately, the historical commit does not really shed any light into why.
> 
>     2. Smaller change: Create a dedicated socket during setup like `UDP_listenerFd` (re-using `setupUDPListeningSocket()`) and call it `UDP_dummyFd` or something, use that and close it during teardown? Would eliminate a lot of code that way, and keep 1 global and it is also closer in implementation to the other socket stuff.
> 
> 
> Both are worth exploring IMO. Getting to option 1 would be awesome. But option 2 seems safer and since this is going into a stable release and will have to be backported to 6X, I would personally lean towards option 2 (with a FIXME comment: "check if we can use UDP_listenerFd to send the dummy message").
> 
> Thoughts?

If we should go to the two routes mentioned, I'm also leaning towards option 2; smaller changes and being able to backport the changes to GP6 would be great.
+1 on the FIXME comment
However, I think that option 1 should be the end goal if it makes things cleaner and more concise.
So overall, both options should be explored...

File: src/backend/cdb/motion/ic_udpifc.c [hunk 3]
```diff
@@ -6890,74 +6895,52 @@ WaitInterconnectQuitUDPIFC(void)
...
-			continue;
-		}
-		break;
-	}
-
-	if (rp == NULL)
-	{
-		elog(LOG, "send dummy packet failed, create socket failed: %m");
+	if (bind(sockfd, (struct sockaddr *) &src_addr, sizeof(src_addr)) < 0) {
+		elog(LOG, "could not bind UDP interconnect socket: %m");
```
[the remark this answers is not in the document]
Comment (Chibin) [reply src/backend/cdb/motion/ic_udpifc.c:6932]: Since we're not really receiving anything, then not really necessary.
I'll remove it.

> Also we are not doing pg_set_noblock() here as before, which we also do for setupUDPListeningSocket.
I'll add it back in.

File: src/backend/cdb/motion/ic_udpifc.c (hunk 1 above)
Comment (Chibin) [reply src/backend/cdb/motion/ic_udpifc.c]: Done!

File: src/backend/cdb/motion/ic_udpifc.c [hunk 4]
```diff
@@ -6977,15 +6961,11 @@ SendDummyPacket(void)
 		goto send_error;
 	}
 
-	pg_freeaddrinfo_all(hint.ai_family, addrs);
 	closesocket(sockfd);
 	return;
```
Comment (soumyadeep2007) [inline src/backend/cdb/motion/ic_udpifc.c:6965]: These 2 lines are now superfluous.

**Review by soumyadeep2007 — COMMENTED** [review]
Things look good as is! Would you want to give option 2 a go though in this PR?

**Comment by Chibin** [comment]
> Things look good as is! Would you want to give option 2 a go though in this PR?

After reading a bit more into the interconnect code, I don't think approach 2 will work.

The term I typically use for Interconnect is vague in the sense that it's all combined together, but within the code. It's separated to two portions (that I know if right now) that interact together.
1st is the Motion Layer IPC (ML) `ml_ipc` and 2nd is the Interconnect.

Motion Layer is what handles the packets and communication of the interconnect layer.
`UDP_listenerFd` happens to be the listening socket of the Motion Layer.

Interconnect is established anytime there's a DML that is being executed.
When an interconnect is established, a new connection will be added to a hash table.
This hash table is used in the motion layer (listening socket UDP_listenerFd) so that ML can facilitate the packet handling, such as ACKing the correct sequence of packets, packet buffering, etc.

So in summary, this commit addresses the issue in the Motion Layer IPC portion, not the "interconnect layer".

To answer option 1:
It's not that we need to use `UDP_listenerFd` socket to send the message, it's that we need a socket to teardown the `UDP_listenerFd`. In other words, we must do `SendDummyPacket` to the `UDP_listenerFd` in order to tear down the Motion Layer IPC faster.

**Review by AJR-VMware — APPROVED** [review]
Assuming a green pipeline, I don't see any problems with the code.
Please do squash down and tidy up the commits before merging.
LGTM!

**Comment by Chibin** [comment]
> Assuming a green pipeline, I don't see any problems with the code. Please do squash down and tidy up the commits before merging. LGTM!

Thank you for reviewing @AJR-VMware !
Yes, I'll be squashing them.
